deps: add support for interface broadcast addresses

pprindeville opened the pull request:
##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

addaleax commented:
Hi! Could you open a pull request against https://github.com/libuv/libuv/ with the libuv changes first? We generally avoid doing our own patches on top of libuv releases. (The libuv part also looks semver-major, which may be a challenge on its own.)

richardlau commented:
There are existing discussions in libuv/libuv#158 and libuv/libuv#1371 about extending libuv's support for network interfaces.

pprindeville commented:
> Hi! Could you open a pull request against https://github.com/libuv/libuv/ with the libuv changes first? We generally avoid doing our own patches on top of libuv releases. (The libuv part also looks semver-major, which may be a challenge on its own.)

Done.  See https://github.com/libuv/libuv/pull/2033.

I wasn't sure if I should branch from `master` or `v1.x`.

pprindeville commented:
> There are existing discussions in libuv/libuv#158 and libuv/libuv#1371 about extending libuv's support for network interfaces.

I saw those, and from what I can tell they deal with larger architectural issues, like getting addresses for non-IP families, interfaces with no layer-3 configuration, etc.  That's a bit orthogonal to what I'm doing here.

pprindeville commented:
I managed to get this (`src/node_os.cc`) to compile against `libuv/master` with my changes, even if there were other files that wouldn't build...  was not able to test run, however.

jasnell commented on doc/api/os.md:
```diff
@@ -259,6 +259,8 @@ The properties available on the assigned network address object include:
 * `address` {string} The assigned IPv4 or IPv6 address
 * `netmask` {string} The IPv4 or IPv6 network mask
+* `broadcast` {string} The IPv4 or IPv6 broadcast address (not available
+  on all interface types)
```
When we say "not available on all interface types"... does that mean it will return `undefined`?

pprindeville commented:
Currently it returns the empty string

gireeshpunathil commented:
how do we move forward on this one?

bnb commented:
heya @pprindeville! any chance you're going to pick this up again? No worries if not.

If not, anyone else?

If not... we should probably close this out.

pprindeville commented:
> heya @pprindeville! any chance you're going to pick this up again? No worries if not.
> 
> If not, anyone else?
> 
> If not... we should probably close this out.

I can try, but I'm not entirely sure what the blocker is.  What do I need to do that I haven't done, or redo, in order to advance this?

bnb commented:
@pprindeville it looks like there's an open question (similar to mine!) in https://github.com/libuv/libuv/pull/2033#pullrequestreview-351825407. I'm assuming that if those can get merged in, this can get merged in if I'm reading the PRs correctly. So, ideally, finishing that PR up then coming back to this one once that lands :)

marco-ippolito commented:
@pprindeville I think the PR that was blocking was merged, or I am wrong?

RedYetiDev commented:
@pprindeville This PR hasn't had any activity for a while (and the [stalebot](https://github.com/nodejs/node/pull/55032) isn't up), so I've marked this as `stalled`. Feel free to continue work, if you still would like to persue this PR.

pprindeville commented:
> @pprindeville This PR hasn't had any activity for a while (and the [stalebot](https://github.com/nodejs/node/pull/55032) isn't up), so I've marked this as `stalled`. Feel free to continue work, if you still would like to persue this PR.

Yeah, after 6 years, whatever expectation of progress I had has dwindled.

pprindeville committed:
os: expose broadcast address in os.networkInterfaces()

Not using the correct broadcast address in applications can cause
disasterous results. Rather than letting the programmer guess at
the broadcast address and try to derive it correctly, allow him to
query the system instead for the correctly configured state.

Fixes: https://github.com/nodejs/node/issues/23437
Depends-on: https://github.com/libuv/libuv/pull/2033
